File: src/LoneWolf/Chapter.hs
{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE DeriveGeneric #-}
{-
---
title: "Solving a gamebook : chapters"
series: Gamebook solver
date: 2017-03-14
---

Chapter modelisation
====================

This article is a literate Haskell representation of the `LoneWolf.Chapter` module at the time of the writing.

It doesn't include any code, only type definitions, and will certainly be the least interesting post of this serie.
It however shows a possible encoding of the chapter's descriptions, which the reader might find interesting on its own.

The last part discusses how the chapter definitions were partially generated.

Extensions and imports
----------------------

Clearly not the best part :)
-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TemplateHaskell #-}

module LoneWolf.Chapter where

import Control.Lens
import Data.Aeson
import Data.Data
import Data.Data.Lens
import qualified Data.Function.Memoize as M
import GHC.Generics (Generic)
import LoneWolf.Character

{-
Basic types
-----------

The following types should get a newtype, but to be honest, it is much more handy that way right now, because of the vicious code generation that is happening later ...
-}

-- each chapter has an identifier, with a value ranging from 1 to 350
type ChapterId = Int

-- some combat effects are timed, for example some fights can be evaded after a certain number of rounds
type Rounds = Int

-- price of stuff, in gold coins
type Price = Int

{-
What's in a chapter?
--------------------

Lone Wolf chapters usually have a descriptive text, some figures, a description of what happens to the payer, and a list of choices. The `title` of the chapter serves no real purpose here, as it is identical (in `String` form) to the associated `ChapterId`.
-}

type Chapter = ChapterG Rational

data ChapterG a = Chapter
  { _title :: String,
    _desc :: String,
    _pchoice :: DecisionG a
  }
  deriving (Show, Eq, Generic, Functor)

instance ToJSON a => ToJSON (ChapterG a) where
  toJSON = genericToJSON jsonOptions

{-
Here is what an actual chapter looks like:

```haskell
 Chapter "346"
         "The driver nods and hands back the ticket. The inn is warm but poorly furnished..."
         (Decisions
          [ ( "Buy a meal, and a room"
            , Conditional
                (HasItem Gold 2)
                (NoDecision (LoseItem Gold 2 (Goto 280))))
          , ( "Just the room"
            , Conditional
                (HasItem Gold 1)
                (NoDecision (LoseItem Gold 1 (MustEat NoHunt (Goto 280)))))
          , ("Nothing", NoDecision (MustEat NoHunt (Goto 280)))
          ])
```

The important part of this model is that the `Decision` type encodes the **player choices**, and contains `ChapterOutcome` fields that encode the **outcome of the choices**.
This is a critical distinction, as it is structural to how the solver works.

The Decision type
-----------------

In details, the `Decision` types has the following constructors:
-}

type Decision = DecisionG Rational

data DecisionG a
  = Decisions [(String, DecisionG a)]
  | RetrieveEquipment (DecisionG a)
  | CanTake Item Int (DecisionG a)
  | Canbuy Item Price (DecisionG a)
  | Cansell Item Price (DecisionG a)
  | Conditional BoolCond (DecisionG a)
  | Special SpecialChapter
  | NoDecision (ChapterOutcomeG a)
  | EvadeFight Rounds ChapterId (FightDetailsG a) (ChapterOutcomeG a)
  | AfterCombat (DecisionG a)
  | RemoveItemFrom Slot Int (DecisionG a) -- must have that least that many items
  | LoseItemFrom Slot Int (DecisionG a) -- always succeeds
  deriving (Show, Eq, Typeable, Data, Generic, Functor)

instance ToJSON a => ToJSON (DecisionG a) where
  toJSON = genericToJSON jsonOptions

data SpecialChapter
  = Cartwheel
  | Portholes
  | B05S127
  | B05S357
  deriving (Show, Eq, Typeable, Data, Generic)

instance ToJSON SpecialChapter where
  toJSON = genericToJSON jsonOptions

{-
 * The `Decisions` constructor represents a list of possible choices for the player, with, for each case, a textual and programmatic description.

 * The `CanTake` and `CanBuy` constructors represent situations where the player can take, or buy, items.
   `CanSell` serves an identical purpose, for selling items.
   These constructors are not necessary, as they could be encoded using other, simpler, primitives. The following are equivalent, with regards to the choices they encode:

    ```haskell
    gaingold = CanTake Gold 10 (NoDecision (Goto 20))

    gaingold' = Decisions
      [ ("Take the gold", NoDecision (GainItem Gold 10 (Goto 20)))
      , ("Take 9 coins", NoDecision (GainItem Gold 9 (Goto 20)))
      , ("Take 8 coins", NoDecision (GainItem Gold 8 (Goto 20)))
      , ...
      , ("Leave the gold", NoDecision (Goto 20)
      ]
    ```

    Having specialized constructors helps a lot. First it makes it easier to write the chapter definitions, and makes them clearer.
    But, more importantly, it will help reducing the computational complexity by deciding locally with specific heuristics.

 * Some decisions are only accessible in certain conditions, for example:

    ```haskell
    (Decisions
     [ ( "If you have the Kai Discipline of Tracking, turn to 13."
         , Conditional (HasDiscipline Tracking) (NoDecision (Goto 13)))
     , ( "If you wish to take the left fork, turn to 155."
         , Conditional
             (Not (HasDiscipline Tracking))
             (NoDecision (Goto 155)))
     , ( "If you wish to take the right fork, turn to 293."
         , Conditional
             (Not (HasDiscipline Tracking))
             (NoDecision (Goto 293)))
     ])
    ```

    In this example, if the player has the `Tracking` discipline, he will be forced to jump to chapter 13. If not, he will be able to choose between chapters 155 and 293.

 * Some chapters are special, and will be handled specifically. More on this later!

 * The `NoDecision` constructor indicates that there is no further choice for the player, and is at the leaf of most `Decision` structures.
   I currently have flagged two chapters as special, but this might evolve in the future.

The ChapterOutcome type
-----------------------

A `ChapterOutcome` describe what happens to the player once he has made a decision.
-}

type ChapterOutcome = ChapterOutcomeG Rational

data ChapterOutcomeG a
  = Fight (FightDetailsG a) (ChapterOutcomeG a)
  | -- | lw loss, equal, lw win
    OneRound (FightDetailsG a) (ChapterOutcomeG a) (ChapterOutcomeG a) (ChapterOutcomeG a)
  | Randomly [(a, ChapterOutcomeG a)]
  | Conditionally [(BoolCond, ChapterOutcomeG a)]
  | Simple [SimpleOutcome] (ChapterOutcomeG a)
  | Goto ChapterId
  | GameLost
  | GameWon
  deriving (Show, Eq, Typeable, Data, Generic, Functor)

instance ToJSON a => ToJSON (ChapterOutcomeG a) where
  toJSON x = case x of
    GameLost -> "GameLost"
    GameWon -> "GameWon"
    Fight a b -> object [("Fight", toJSON (a, b))]
    Simple a b -> object [("Simple", toJSON (a, b))]
    Conditionally l -> object [("Conditionally", toJSON l)]
    Randomly l -> object [("Randomly", toJSON l)]
    Goto d -> object [("Goto", toJSON d)]
    OneRound a b c d -> object [("OneRound", toJSON (a, b, c, d))]

data SimpleOutcome
  = DamagePlayer Endurance
  | HealPlayer Endurance
  | FullHeal
  | HalfHeal
  | GainItem Item Int
  | LoseItem Item Int
  | LoseItemKind [Slot]
  | MustEat CanHunt
  | StoreEquipment -- drop all equipment, saving it in case we find it again
  | SetFlag Flag
  | ClearFlag Flag
  deriving (Show, Eq, Typeable, Data, Generic)

instance ToJSON SimpleOutcome where
  toJSON x = case x of
    DamagePlayer a -> object [("DamagePlayer", toJSON a)]
    HealPlayer a -> object [("HealPlayer", toJSON a)]
    LoseItemKind a -> object [("LoseItemKind", toJSON a)]
    MustEat a -> object [("MustEat", toJSON a)]
    SetFlag a -> object [("SetFlag", toJSON a)]
    ClearFlag a -> object [("ClearFlag", toJSON a)]
    FullHeal -> "FullHeal"
    HalfHeal -> "HalfHeal"
    StoreEquipment -> "StoreEquipment"
    GainItem a b -> object [("GainItem", toJSON (a, b))]
    LoseItem a b -> object [("LoseItem", toJSON (a, b))]

hasCombat :: ChapterOutcome -> Bool
hasCombat o =
  case o of
    Fight _ _ -> True
    OneRound {} -> True
    Randomly probs -> any (hasCombat . snd) probs
    Conditionally choices -> any (hasCombat . snd) choices
    Simple _ o' -> hasCombat o'
    Goto _ -> False
    GameLost -> False
    GameWon -> False

{-
A constructor of interest is the `MustEat` constructor.
It represents chapters where the player loses some hit points if he doesn't have a meal in his backpack.
In the Lone Wolf series, the `Hunting` discipline let the player ignore these events.
But in this specific book, there are a couple of chapters where `Hunting` can't be used!
In a previous tentative, I used a flag in the game state to indicate whether or not hunting was permitted (it's explicitely told when it stops and when it starts).
In general, it is better to make the game state as simple as possible, so as to get good memoization performance, so I decided to manually track the chapters where hunting was disabled, and mark them as such.

```haskell
Conditionally
 [ (HasItem Laumspur 1, Goto 145)
 , (HasDiscipline Healing, Goto 210)
 , (Always True, Goto 275)
 ]
```

Note how it works like Haskell's guards, and ends up with the equivalent of `otherwise`, which I called ... `botherwise`:
-}

botherwise :: BoolCond
botherwise = Always True

data CanHunt = Hunt | NoHunt
  deriving (Show, Eq, Typeable, Data, Generic)

instance ToJSON CanHunt where
  toJSON = genericToJSON jsonOptions

data BoolCond
  = HasDiscipline Discipline
  | Not BoolCond
  | COr BoolCond BoolCond
  | CAnd BoolCond BoolCond
  | HasItem Item Int
  | Always Bool
  | HasEndurance Endurance
  | HasFlag Flag
  | HasLevel KaiLevel
  deriving (Show, Eq, Typeable, Data, Generic)

instance ToJSON BoolCond where
  toJSON = genericToJSON jsonOptions

(.&&.) :: BoolCond -> BoolCond -> BoolCond
(.&&.) = CAnd

{-
The boolean conditions type (`BoolCond`) is pretty self-explanatory.

Combat description
------------------

The three main parts of a combat descriptions are the name of the opponent, its combat skill, and its endurance.
There are however many combat situations that involve special modifiers which are detailled here.
Fights against consecutive opponents are handled by chaining the `Fight` and `EvadeFight` constructors.
-}

type FightDetails = FightDetailsG Rational

data FightDetailsG a = FightDetails
  { _opponent :: String,
    _fcombatSkill :: CombatSkill,
    _fendurance :: Endurance,
    _fightMod :: [FightModifierG a]
  }
  deriving (Show, Eq, Typeable, Data, Generic, Functor)

instance ToJSON a => ToJSON (FightDetailsG a) where
  toJSON = genericToJSON jsonOptions

type FightModifier = FightModifierG Rational

data FightModifierG a
  = Undead
  | MindblastImmune
  | Timed Int (FightModifierG a)
  | CombatBonus CombatSkill
  | BareHanded
  | FakeFight ChapterId
  | EnemyMindblast
  | ForceEMindblast -- forced loss of 2 pts/turn
  | PlayerInvulnerable
  | DoubleDamage -- chapter 306
  | Evaded ChapterId
  | OnDamage ChapterId
  | OnNotYetWon ChapterId -- if won in less than that amount of rounds
  | MultiFight -- set when this is only part of a single chapter fight
  | EnemyInvulnerable -- rounds during which the enemy is invulnerable
  | OnLose ChapterId -- if the fight is lost, do not die, but jump
  | StopFight ChapterId -- immediately stop the fight and jump
  | DPR Endurance -- damage per round inflicted to the opponent
  | NoPotion -- can't take a potion for this fight
  | Poisonous a -- opponent doesn't damage, but can instakill on damage (cf b03s088)
  deriving (Show, Eq, Typeable, Data, Generic, Ord, Functor)

instance ToJSON a => ToJSON (FightModifierG a) where
  toJSON x = case x of
    Undead -> String "Undead"
    MindblastImmune -> String "MindblastImmune"
    Timed t m -> object [("Timed", toJSON (t, m))]
    CombatBonus a -> object [("CombatBonus", toJSON a)]
    BareHanded -> String "BareHanded"
    FakeFight a -> object [("FakeFight", toJSON a)]
    EnemyMindblast -> String "EnemyMindblast"
    ForceEMindblast -> String "ForceEMindblast"
    PlayerInvulnerable -> String "PlayerInvulnerable"
    DoubleDamage -> String "DoubleDamage"
    Evaded a -> object [("Evaded", toJSON a)]
    OnDamage a -> object [("OnDamage", toJSON a)]
    OnNotYetWon a -> object [("OnNotYetWon", toJSON a)]
    MultiFight -> String "MultiFight"
    EnemyInvulnerable -> String "EnemyInvulnerable"
    OnLose a -> object [("OnLose", toJSON a)]
    StopFight a -> object [("StopFight", toJSON a)]
    DPR a -> object [("DPR", toJSON a)]
    NoPotion -> String "NoPotion"
    Poisonous a -> object [("Poisonous", toJSON a)]

data KaiLevel
  = Novice
  | Intuite
  | Doan
  | Acolyte
  | Initiate
  | Aspirant
  | Guardian
  | Warmarn
  | Savant
  | Master
  deriving (Show, Eq, Ord, Enum, Bounded, Data, Generic)

instance ToJSON KaiLevel where
  toJSON = genericToJSON jsonOptions

getLevel :: CharacterConstant -> KaiLevel
getLevel ccst = case length (_discipline ccst) of
  1 -> Novice
  2 -> Intuite
  3 -> Doan
  4 -> Acolyte
  5 -> Initiate
  6 -> Aspirant
  7 -> Guardian
  8 -> Warmarn
  9 -> Savant
  10 -> Master
  _ -> error ("invalid amount of disciplines: " ++ show (_discipline ccst))

{-
 * `Undead`: undead creatures take double damage from the `Sommerswerd`.
 * `MindblastImmune`: the opponent is not affected by the `MindBlast` ability of the player.
 * `CombatBonus`: the player receives a combat bonus for this fight.
 * `BareHanded`: the player is forced to fight bare handed, reducing his combat skill by 4 points.
 * `FakeFight`: this isn't a real fight, and the player will have his endurance restored at the end of the fight. The `ChapterId` is provided as the destination if he loses the fight. This happens in [chapter 276](https://www.projectaon.org/en/xhtml/lw/02fotw/sect276.htm).
 * `EnemyMindblast`: the enemy has the `MindBlast` ability!
 * `PlayerInvulnerable` the player cannot be harmed.
 * `DoubleDamage`: happens in [chapter 306](https://www.projectaon.org/en/xhtml/lw/02fotw/sect306.htm), where the player inflicts double damage.
 * `Timed`: the effect only lasts a given number of rounds.
-}

jsonOptions :: Options
jsonOptions =
  defaultOptions
    { fieldLabelModifier = dropWhile (== '_'),
      allNullaryToStringTag = True,
      unwrapUnaryRecords = True,
      tagSingleConstructors = False,
      sumEncoding = ObjectWithSingleField
    }

{-
Lenses, plates and utilities
----------------------------
-}

makePrisms ''ChapterOutcomeG
makePrisms ''SimpleOutcome
makePrisms ''DecisionG
makePrisms ''FightModifierG
makeLenses ''FightDetailsG
makeLenses ''ChapterG

moneyCond :: Int -> ChapterOutcome -> Decision
moneyCond price = Conditional (HasItem Gold price) . NoDecision . Simple [LoseItem Gold price]

outcomePlate :: Traversal' Decision ChapterOutcome
outcomePlate = biplate

_Outcome :: Traversal' Decision ChapterOutcome
_Outcome = outcomePlate

{-
All of these are mainly useful in the `LoneWolf.XML` module.

The actual chapters data
========================

The chapters are described in the `LoneWolf.Book02` module, but it's all generated code!
The `LoneWolf.XML` module contains all the code that is necessary to parse the XML file provided by project Aon, compute a baseline, adapt all chapters that have special rules and generate code.

In previous iterations of this project, I used to manually generate all the chapter's definition, and chapter 346, that could be seen at the beginning of this article, would look like that:

```haskell
(346, UnsetFlag CanHunt
    ( Decisions [ ("buy meal", ItemPay Gold 1 (Outcome (Goto 1104)))
                , ("eat own meal", ItemPay Meal 1 (Outcome (Goto 1104)))
                , ("starve", Outcome (ODamage 3 (Goto 1104))) ] ))
```

Compared to the current version, the chapter's text is missing, and the choice descriptions are not as nice. It was also harder to debug, and one can see the `UnsetFlag CanHunt` situation I mentioned earlier.

The current situation is a net gain, as it has all the original text content, and also wrote most of the code for me. It was however still a **lot** of work to read all the chapters and make sure they were properly encoded.
I also realized my previous encoding was imperfect in parts (I never realized there was a chapter where you could sell your weapons!), and I have no illusions there are mistakes in the current definitions.

Next time will have actual code, with a rules interpreter!

-}

M.deriveMemoizable ''FightModifierG

File: gamebooksolver-base/src/lonewolf/chapter.rs
use crate::solver::rational::Rational;

use super::mini::{Discipline, Flag, Item, Slot};
use serde::{Deserialize, Serialize};
use std::{borrow::Cow, hash::Hash};

#[derive(PartialEq, Eq, Debug, Deserialize, Clone)]
pub struct Chapter<P> {
    pub title: String,
    pub desc: String,
    pub pchoice: Decision<P>,
}

impl<P: Clone> Chapter<P> {
    pub fn map_proba<F, P2>(self, f: &F) -> Chapter<P2>
    where
        F: Fn(&P) -> P2,
    {
        Chapter {
            title: self.title,
            desc: self.desc,
            pchoice: self.pchoice.map_proba(f),
        }
    }
}

#[derive(PartialEq, Eq, Debug, Deserialize, Hash, Clone, Copy, PartialOrd, Ord, Serialize)]
pub struct Rounds(pub u8);
#[derive(PartialEq, Eq, Debug, Deserialize, Hash, Clone, Copy, PartialOrd, Ord, Serialize)]
pub struct Price(pub u8);
#[derive(PartialEq, Eq, Debug, Deserialize, Hash, Clone, Copy, PartialOrd, Ord, Serialize)]
pub struct ChapterId(pub u16);
#[derive(PartialEq, Eq, Debug, Deserialize, Hash, Clone, Copy, PartialOrd, Ord, Serialize)]
pub struct Endurance(pub i8);
#[derive(PartialEq, Eq, Debug, Deserialize, Hash, Clone, Copy, PartialOrd, Ord, Serialize)]
pub struct CombatSkill(pub i8);

impl Endurance {
    pub fn damage(&self, v: u8) -> Self {
        Endurance(std::cmp::max(self.0 - v as i8, 0))
    }

    pub const fn dead(&self) -> bool {
        self.0 <= 0
    }
}

#[derive(PartialEq, Eq, Debug, Deserialize, Clone)]
pub enum Decision<P> {
    Decisions(Vec<(String, Decision<P>)>),
    RetrieveEquipment(Box<Decision<P>>),
    CanTake(Item, u8, Box<Decision<P>>),
    Canbuy(Item, Price, Box<Decision<P>>),
    Cansell(Item, Price, Box<Decision<P>>),
    Conditional(BoolCond, Box<Decision<P>>),
    Special(SpecialChapter),
    #[serde(rename = "NoDecision")]
    None(ChapterOutcome<P>),
    EvadeFight(Rounds, ChapterId, FightDetails<P>, ChapterOutcome<P>),
    AfterCombat(Box<Decision<P>>),
    RemoveItemFrom(Slot, u8, Box<Decision<P>>),
    LoseItemFrom(Slot, u8, Box<Decision<P>>),
}

pub struct OutcomeIt<'t, P> {
    decs: Vec<&'t Decision<P>>,
    oc: Vec<&'t ChapterOutcome<P>>,
}

impl<'t, P: Rational> Iterator for OutcomeIt<'t, P> {
    type Item = Cow<'t, ChapterOutcome<P>>;

    fn next(&mut self) -> Option<Self::Item> {
        match self.oc.pop() {
            None => match self.decs.pop()? {
                Decision::Decisions(items) => {
                    self.decs.extend(items.iter().rev().map(|x| &x.1));
                    self.next()
                }
                Decision::Canbuy(_, _, decision)
                | Decision::Cansell(_, _, decision)
                | Decision::Conditional(_, decision)
                | Decision::AfterCombat(decision)
                | Decision::RemoveItemFrom(_, _, decision)
                | Decision::LoseItemFrom(_, _, decision)
                | Decision::CanTake(_, _, decision)
                | Decision::RetrieveEquipment(decision) => {
                    self.decs.push(decision);
                    self.next()
                }
                Decision::Special(special_chapter) => match special_chapter {
                    SpecialChapter::Cartwheel => Some(Cow::Owned(ChapterOutcome::Randomly(vec![
                        (P::f_f64(0.5), ChapterOutcome::Goto(ChapterId(169))),
                        (P::f_f64(0.5), ChapterOutcome::Goto(ChapterId(186))),
                    ]))),
                    SpecialChapter::Portholes => Some(Cow::Owned(ChapterOutcome::Goto(ChapterId(197)))),
                    SpecialChapter::B05S127 => Some(Cow::Owned(ChapterOutcome::Randomly(vec![
                        (P::f_f64(0.5), ChapterOutcome::Goto(ChapterId(159))),
                        (P::f_f64(0.5), ChapterOutcome::Goto(ChapterId(93))),
                    ]))),
                    SpecialChapter::B05S357 => Some(Cow::Owned(ChapterOutcome::Randomly(vec![
                        (P::f_f64(0.5), ChapterOutcome::Goto(ChapterId(207))),
                        (P::f_f64(0.5), ChapterOutcome::Goto(ChapterId(224))),
                    ]))),
                },
                Decision::None(chapter_outcome) | Decision::EvadeFight(_, _, _, chapter_outcome) => {
                    Some(Cow::Borrowed(chapter_outcome))
                }
            },
            Some(o) => Some(Cow::Borrowed(o)),
        }
    }
}

impl<P: Clone> Decision<P> {
    pub fn iter_outcomes<'t>(&'t self) -> OutcomeIt<'t, P> {
        OutcomeIt {
            decs: vec![self],
            oc: Vec::new(),
        }
    }

    pub fn map_proba<F, P2>(self, f: &F) -> Decision<P2>
    where
        F: Fn(&P) -> P2,
    {
        use Decision::*;
        let convert = |b: Box<Decision<P>>| -> Box<Decision<P2>> {
            Box::new({
                let b2: &Decision<P> = &b;
                b2.clone().map_proba(f)
            })
        };
        match self {
            Decisions(v) => Decisions(v.into_iter().map(|(t, sub)| (t, sub.map_proba(f))).collect()),
            RetrieveEquipment(nxt) => RetrieveEquipment(convert(nxt)),
            CanTake(i, q, nxt) => CanTake(i, q, convert(nxt)),
            Canbuy(i, p, nxt) => Canbuy(i, p, convert(nxt)),
            Cansell(i, p, nxt) => Cansell(i, p, convert(nxt)),
            Conditional(bc, nxt) => Conditional(bc, convert(nxt)),
            Special(s) => Special(s),
            None(co) => None(co.map_proba(f)),
            EvadeFight(r, cid, fd, co) => EvadeFight(r, cid, fd.map_proba(f), co.map_proba(f)),
            AfterCombat(nxt) => AfterCombat(convert(nxt)),
            RemoveItemFrom(sl, q, nxt) => RemoveItemFrom(sl, q, convert(nxt)),
            LoseItemFrom(sl, q, nxt) => LoseItemFrom(sl, q, convert(nxt)),
        }
    }
}

#[derive(PartialEq, Eq, Debug, Deserialize, Clone, Copy)]
pub enum SpecialChapter {
    Cartwheel,
    Portholes,
    B05S127,
    B05S357,
}

#[derive(PartialEq, Eq, Debug, Deserialize, Clone)]
pub enum ChapterOutcome<P> {
    Fight(FightDetails<P>, Box<ChapterOutcome<P>>),
    OneRound(
        FightDetails<P>,
        Box<ChapterOutcome<P>>,
        Box<ChapterOutcome<P>>,
        Box<ChapterOutcome<P>>,
    ),
    Randomly(Vec<(P, ChapterOutcome<P>)>),
    Conditionally(Vec<(BoolCond, ChapterOutcome<P>)>),
    Simple(Vec<SimpleOutcome>, Box<ChapterOutcome<P>>),
    Goto(ChapterId),
    GameLost,
    GameWon,
}

pub struct SimpleOutcomeIt<'t, P> {
    si: Vec<&'t SimpleOutcome>,
    co: Vec<&'t ChapterOutcome<P>>,
}

impl<'t, P> Iterator for SimpleOutcomeIt<'t, P> {
    type Item = &'t SimpleOutcome;

    fn next(&mut self) -> Option<Self::Item> {
        match self.si.pop() {
            Some(x) => Some(x),
            None => match self.co.pop()? {
                ChapterOutcome::OneRound(_, co1, co2, co3) => {
                    self.co.push(co3);
                    self.co.push(co2);
                    self.co.push(co1);
                    self.next()
                }
                ChapterOutcome::Fight(_, chapter_outcome) => {
                    self.co.push(chapter_outcome);
                    self.next()
                }
                ChapterOutcome::Randomly(items) => {
                    self.co.extend(items.iter().map(|x| &x.1));
                    self.next()
                }
                ChapterOutcome::Conditionally(items) => {
                    self.co.extend(items.iter().map(|x| &x.1));
                    self.next()
                }
                ChapterOutcome::Simple(simple_outcomes, chapter_outcome) => {
                    self.co.push(chapter_outcome);
                    self.si.extend(simple_outcomes);
                    self.next()
                }
                ChapterOutcome::Goto(_) | ChapterOutcome::GameWon | ChapterOutcome::GameLost => self.next(),
            },
        }
    }
}

impl<P> ChapterOutcome<P> {
    pub fn simple_outcomes(&'_ self) -> SimpleOutcomeIt<'_, P> {
        SimpleOutcomeIt {
            si: Vec::new(),
            co: vec![self],
        }
    }
}

impl<P: Clone> ChapterOutcome<P> {
    pub fn map_proba<F, P2>(self, f: &F) -> ChapterOutcome<P2>
    where
        F: Fn(&P) -> P2,
    {
        let convert = |b: Box<ChapterOutcome<P>>| -> Box<ChapterOutcome<P2>> {
            Box::new({
                let b2: &ChapterOutcome<P> = &b;
                b2.clone().map_proba(f)
            })
        };
        use ChapterOutcome::*;
        match self {
            Fight(fd, co) => Fight(fd.map_proba(f), convert(co)),
            OneRound(fd, cl, ce, cw) => OneRound(fd.map_proba(f), convert(cl), convert(ce), convert(cw)),
            Randomly(lst) => Randomly(lst.into_iter().map(|(p, co2)| (f(&p), co2.map_proba(f))).collect()),
            Conditionally(lst) => Conditionally(lst.into_iter().map(|(c, co2)| (c, co2.map_proba(f))).collect()),
            Simple(sos, nxt) => Simple(sos, convert(nxt)),
            Goto(cid) => Goto(cid),
            GameLost => GameLost,
            GameWon => GameWon,
        }
    }
}

#[derive(PartialEq, Eq, Debug, Deserialize, Clone)]
pub enum SimpleOutcome {
    DamagePlayer(Endurance),
    HealPlayer(Endurance),
    FullHeal,
    HalfHeal,
    GainItem(Item, u8),
    LoseItem(Item, u8),
    LoseItemKind(Vec<Slot>),
    MustEat(CanHunt),
    StoreEquipment,
    SetFlag(Flag),
    ClearFlag(Flag),
}

#[derive(PartialEq, Eq, Debug, Deserialize, Clone, Copy)]
pub enum CanHunt {
    Hunt,
    NoHunt,
}

#[derive(PartialEq, Eq, Debug, Deserialize, Clone)]
pub enum BoolCond {
    HasDiscipline(Discipline),
    Not(Box<BoolCond>),
    COr(Box<BoolCond>, Box<BoolCond>),
    CAnd(Box<BoolCond>, Box<BoolCond>),
    HasItem(Item, u8),
    Always(bool),
    HasEndurance(Endurance),
    HasFlag(Flag),
    HasLevel(KaiLevel),
}

#[derive(PartialEq, Eq, Debug, Deserialize, Clone)]
pub struct FightDetails<P> {
    pub opponent: String,
    #[serde(rename = "fcombatSkill")]
    pub combat_skill: CombatSkill,
    #[serde(rename = "fendurance")]
    pub endurance: Endurance,
    #[serde(rename = "fightMod")]
    pub fight_mod: Vec<FightModifier<P>>,
}

impl<P: Clone> FightDetails<P> {
    pub fn map_proba<F, P2>(self, f: &F) -> FightDetails<P2>
    where
        F: Fn(&P) -> P2,
    {
        FightDetails {
            opponent: self.opponent,
            combat_skill: self.combat_skill,
            endurance: self.endurance,
            fight_mod: self.fight_mod.into_iter().map(|m| m.map_proba(f)).collect(),
        }
    }
}

#[derive(PartialEq, Eq, Debug, Deserialize, Hash, Clone, PartialOrd, Ord)]
pub enum FightModifier<P> {
    Undead,
    MindblastImmune,
    Timed(u8, Box<FightModifier<P>>),
    CombatBonus(CombatSkill),
    BareHanded,
    FakeFight(ChapterId),
    EnemyMindblast,
    ForceEMindblast,
    PlayerInvulnerable,
    DoubleDamage,
    Evaded(ChapterId),
    OnDamage(ChapterId),
    OnNotYetWon(ChapterId),
    MultiFight,
    EnemyInvulnerable,
    OnLose(ChapterId),
    StopFight(ChapterId),
    #[serde(rename = "DPR")]
    Dpr(Endurance),
    NoPotion,
    Poisonous(P),
}

impl<P: Clone> FightModifier<P> {
    pub fn map_proba<F, P2>(self, f: &F) -> FightModifier<P2>
    where
        F: Fn(&P) -> P2,
    {
        use FightModifier::*;
        match self {
            Undead => Undead,
            MindblastImmune => MindblastImmune,
            Timed(t, nxt) => Timed(t, {
                let nxt2: &FightModifier<P> = &nxt;
                Box::new(nxt2.clone().map_proba(f))
            }),
            CombatBonus(cs) => CombatBonus(cs),
            BareHanded => BareHanded,
            FakeFight(cid) => FakeFight(cid),
            EnemyMindblast => EnemyMindblast,
            ForceEMindblast => ForceEMindblast,
            PlayerInvulnerable => PlayerInvulnerable,
            DoubleDamage => DoubleDamage,
            Evaded(cid) => Evaded(cid),
            OnDamage(cid) => OnDamage(cid),
            OnNotYetWon(cid) => OnNotYetWon(cid),
            MultiFight => MultiFight,
            EnemyInvulnerable => EnemyInvulnerable,
            OnLose(cid) => OnLose(cid),
            StopFight(cid) => StopFight(cid),
            Dpr(e) => Dpr(e),
            NoPotion => NoPotion,
            Poisonous(p) => Poisonous(f(&p)),
        }
    }
}

#[derive(PartialEq, Eq, Debug, PartialOrd, Ord, Deserialize, Clone, Copy)]
#[repr(u8)]
pub enum KaiLevel {
    Novice = 1,
    Intuite = 2,
    Doan = 3,
    Acolyte = 4,
    Initiate = 5,
    Aspirant = 6,
    Guardian = 7,
    Warmarn = 8,
    Savant = 9,
    Master = 10,
}
